### PR TITLE
SHA-3 performance & code size

### LINKED ISSUE #8819: Bad SHA3 performance with GCC
It seems that GCC is pretty bad at optimizing our SHA3 code: it's between 2× and 3× slower than Clang. Here's `programs/test/benchmark sha3_256` built with `gcc-11 -O3` on my Linux/x86_64 computer:
```
  SHA3-256                 :     118404 KiB/s,         22 cycles/byte
```

Compare with `clang-14 -O3` on the same machine:
```
  SHA3-256                 :     241472 KiB/s,         10 cycles/byte
```

There's a similar ratio with `-m32`. @dave-rodgman reports a similar difference on armv8 M1. Dave looked into it a bit and found that clang doesn't autovectorize `keccak_f1600` whereas GCC does a bit, but turning off auto-vectorization only improves by 10% so that's not the main cause.

With other algorithms, Clang and GCC are within 5% or so of each other. So there seems to be something specific in our SHA3 that GCC doesn't optimize well.

### PR DESCRIPTION
**Pull request by daverodgman**
## Description

Improve SHA-3 performance and code-size.

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** not required - no functional change
- [x] **backport** not required
- [x] **tests** not required

Fixes #8819

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
On my x86_64 PC, for `programs/test/benchmark sha3_256`, cfb126f1baa05c204e7cc5f378f9fdf2c295711b brings `gcc-12 -O3` up to par with `clang-14 -O3` for x86_64, and `gcc-12 -m32 -O3` is about 10% slower than `clang-14 -m32 -O3`. No measurable performance hit for Clang.

**Comment by daverodgman** [comment]
Relative to baseline, I see:

|FIELD1  |no unroll|unroll pi|unroll rho|unroll both|
|--------|---------|---------|----------|-----------|
|Os gcc  |-8%      |9%       |2%        |25%        |
|Os clang|8%       |20%      |12%       |20%        |
|O1 gcc  |-12%     |2%       |7%        |23%        |
|O1 clang|10%      |29%      |2%        |20%        |
|O2 gcc  |-3%      |9%       |7%        |39%        |
|O2 clang|-30%     |-31%     |-4%       |-5%        |
|O3 gcc  |218%     |212%     |214%      |213%       |
|O3 clang|2%       |0%       |1%        |0%         |

i.e., we probably want to forcibly unroll the `rho` inner loop to avoid a large regression for `clang -O2`.

**Comment by daverodgman** [comment]
There's an additional ~20% perf from using 64-bit tables for rho and pi, but I suspect this wouldn't help 32-bit platforms and it adds several hundred bytes of code size, so it feels like the wrong direction to go.

**Comment by daverodgman** [comment]
Additionally, rolling up the chi loop saves 448 bytes, but reduces performance moderately everywhere (improvements reduced by around 0 - 14%, but still better than baseline), except for `gcc -O2` which ends up 38% worse than baseline. So probably worthwhile for clang but not gcc IMO.

With this change, 865480279c shows size down by 946 bytes on clang, and by 48 bytes on gcc, and performance is (just about) same-or-better for everything.

| -     | -  | 8fe2e36de | 6fd6542e9c | 6fd6 change | 865480279c | 8654 change |
|-------|----|-----------|------------|-------------|------------|-------------|
| gcc   | Os | 151427    | 196954     | 30%         | 196860     | 30%         |
| gcc   | O1 | 161104    | 180099     | 12%         | 177839     | 10%         |
| gcc   | O2 | 163112    | 211671     | 30%         | 210928     | 29%         |
| gcc   | O3 | 141225    | 440946     | 212%        | 439819     | 211%        |
| clang | Os | 149974    | 168726     | 13%         | 149985     | 0%          |
| clang | O1 | 138158    | 174211     | 26%         | 161635     | 17%         |
| clang | O2 | 293794    | 297117     | 1%          | 289739     | -1%         |
| clang | O3 | 397856    | 376145     | -5%         | 407140     | 2%          |

**Review by minosgalanakis — APPROVED** [review]
LGTM

**Comment by daverodgman** [comment]
On x86_64, @minosgalanakis reports a 19% perf regression for `gcc -Os`; 202% uplift for `gcc -O3`; everything else -2% to +10%. The results seemed a bit variable - may not be very reliable numbers.

**Comment by daverodgman** [comment]
@davidhorstmann-arm did you manage to get some numbers for x86-64?

**Comment by davidhorstmann-arm** [comment]
@daverodgman Apologies, I didn't get round to it. I'm running them now though

**Comment by davidhorstmann-arm** [comment]
These are the numbers I have from x86_64:
| | | 8fe2e36de | 6fd6542e9c | improvement |
|-----------|------------|-------------|--------|------|
| gcc       | Os         | 140447      | 107951 | -23 %  |
| gcc       | O1         | 130248      | 169515 | 30%  |
| gcc       | O2         | 140722      | 182219 | 30%  |
| gcc       | O3         | 134384      | 308645 | 230% |
| clang     | Os         | 112573      | 189634 | 68%  |
| clang     | O1         | 153903      | 143763 | -6.5%  |
| clang     | O2         | 130712      | 335037 | 156%   |
| clang     | O3         | 325279      | 324442 | -0.3%  |

File: library/sha3.c [hunk 1]
```diff
@@ -63,39 +84,98 @@ static void keccak_f1600(mbedtls_sha3_context *ctx)
...
         t = s[1];
-        for (i = 0; i < 24; i++) {
-            SWAP(s[pi[i]], t);
+#if !defined(MBEDTLS_SHA3_PI_UNROLL)
+        for (i = 0; i < 24; i += 4) {
+            uint32_t p = pi[i >> 2];
+            for (unsigned j = 0; j < 4; j++) {
+                uint8_t p8 = (uint8_t) (p >> 24);
+                p <<= 8;
+                SWAP(s[p8], t);
```
Comment (gilles-peskine-arm) [inline library/sha3.c:143]: Minor: why not this, with the endianness of `pi` reversed? It feels simpler, and at a guess it has the same instruction count and performance.
```suggestion
                SWAP(s[p & 0xff], t);
                p >>= 8;
```

File: library/sha3.c [hunk 2]
```diff
@@ -63,39 +84,98 @@ static void keccak_f1600(mbedtls_sha3_context *ctx)
...
+            }
         }
+#else
+        uint32_t p = pi[0];
+        SWAP(s[MBEDTLS_BYTE_3(p)], t); SWAP(s[MBEDTLS_BYTE_2(p)], t);
+        SWAP(s[MBEDTLS_BYTE_1(p)], t); SWAP(s[MBEDTLS_BYTE_0(p)], t);
```
Comment (gilles-peskine-arm) [inline library/sha3.c:149]: Minor: I'd find this more readable with the intermediate variable `p`.
```suggestion
        SWAP(s[MBEDTLS_BYTE_3(pi[0])], t); SWAP(s[MBEDTLS_BYTE_2(pi[0])], t);
        SWAP(s[MBEDTLS_BYTE_1(pi[0])], t); SWAP(s[MBEDTLS_BYTE_0(pi[0])], t);
```
etc.

File: library/sha3.c [hunk 3]
```diff
@@ -14,6 +14,30 @@
...
+ * from manually unrolling at higher optimisation levels.
+ *
+ * Rolling up the theta loop saves a lot of code-size at small performance cost. The code-size
+ * saving then enables us to unroll the other loops for a net code-size saving with a net
+ * performance win.
+ *
+ * Depending on your compiler and target, it may be beneficial to adjust these; the defaults here
+ * should give sensible trade-offs for gcc and clang.
+ */
+#undef   MBEDTLS_SHA3_THETA_UNROLL //no-check-names
```
Comment (gilles-peskine-arm) [inline library/sha3.c:30]: You can avoid the need for `//no-check-names` by giving the `.c`-file-internal macros a name that doesn't have with `MBEDTLS_`. That's just for information, I have no particular preference one way or the other.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I've reviewed the code but not done any benchmarks. The code looks correct to me. However, I am opposed to having code blocks that are not tested on the CI (apart from platform support limitations). Please remove any blocks that are never enabled, and ensure that all remaining blocks are enabled by at least one `all.sh` component.

File: library/sha3.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (daverodgman) [reply library/sha3.c:30]: This is a tricky one. I agree re testing. But the "dead" code is useful - different compilers / architectures can have different trade-offs and people might reasonably want something different for their use-case, and this gives them an easy way to do it (although I don't think they need to be proper options - I'm OK with people having to carry a patch for this level of fine-tuning). We can't really have a one-size-fits-all best answer.

At the moment, I'm still not totally sure what the right set of trade-offs is, so this is useful for experimenting.

I think the right answer may be to keep the four options (with a sensible default), and test all-enabled / all-disabled.

File: library/sha3.c (hunk 2 above)
Comment (daverodgman) [reply library/sha3.c:149]: For me, this is less readable.

**Comment by daverodgman** [comment]
I've pushed 40c837dc40592 but github isn't showing it at the moment

**Review by gilles-peskine-arm — DISMISSED** [review]
LGTM

I think that's a good way of adding low-pain configuration options: they are strictly internal to a single module and don't affect the API or ABI, so the burden of testing them is fully self-contained. This could be inspiration for 4.0: https://github.com/Mbed-TLS/mbedtls/issues/8870

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM - thanks

**Comment by tom-cosgrove-arm** [comment]
Not merging yet to allow @davidhorstmann-arm to gather some more performance data

**Comment by daverodgman** [comment]
@davidhorstmann-arm numbers for x86-64 with show:

Size at worst 16 bytes bigger (gcc and clang -O3), everything else 176 - 805 bytes better.

Perf 23% worse for gcc -Os; -2% gcc -O3, 0% clang -O3; everything else 30% - 75% faster

**Review by tom-cosgrove-arm — DISMISSED** [review]
LGTM

**Comment by daverodgman** [comment]
After further review w.r.t. data from David, I've used the same set of defaults for -O2 and -O3, and adjusted -Os. The conclusions and data above are correct, but the PR didn't reflect the intended settings. All comparison data in this comment is relative to development.

For -O3, for all settings, there are no significant perf regressions (worst was 15%), and we don't care about size. It's also much less common to use this than O2. So I haven't considered -O3 in selecting the options, and focused on O2 and Os only.

For O2, no change to what the previous iteration of this PR intended. All platforms have same-or-better perf (0 - 146% improvement). Size improvement is 96, 120 bytes on aarch64, and 256, 520 on x86-64 (for clang, gcc respectively).

For Os, this is different to what the previous iteration intended, to focus more on size at a slight perf cost (this is probably what users want when optimising for size). Size improvement is 232, 220 bytes on aarch64, and 426,874 on x86-64. Perf is -7%, 16% on aarch64 and 48%,-31% on gcc.

I've given up on trying to ensure that gcc perf closely resembles clang perf. Due to different compiler choices regarding unrolling, this is very hard to assure and probably the wrong thing to do.

In table form (perf delta is higher-is-better, size delta is lower-is-better):

// old and current version of PR
arch | setting | compiler | optimisation level | perf delta | size delta
arm | 13 PtCR | clang | O3 | -7.8% | -12
x86 | 13 PtCR | clang | O3 | -2.2% | -32
arm | 13 PtCR | clang | O2 | 0.4% | -96
arm | 13 PtCR | gcc | O2 | 26.4% | -120
x86 | 13 PtCR | gcc | O2 | 27.9% | -520
x86 | 13 PtCR | gcc | O3 | 129.8% | -32
x86 | 13 PtCR | clang | O2 | 146.2% | -256
arm | 13 PtCR | gcc | O3 | 207.0% | -536

// current version of PR
x86 |  9 PtcR | gcc | Os | -30.9% | -426
arm |  9 PtcR | clang | Os | -7.2% | -232
arm |  9 PtcR | gcc | Os | 16.7% | -220
x86 |  9 PtcR | clang | Os | 48.1% | -874

// old version of PR
x86 | 10 pTcR | gcc | Os | -1.6% | -360
arm | 10 pTcR | clang | Os | 2.2% | -76
arm | 10 pTcR | gcc | Os | 5.3% | -176
x86 | 10 pTcR | clang | Os | 21.1% | -572

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Pull request merged**